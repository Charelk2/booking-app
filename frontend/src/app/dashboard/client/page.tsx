"use client";

import React, { useEffect, useState, useMemo } from "react";
import { useRouter, usePathname, useSearchParams } from "next/navigation";
import MainLayout from "@/components/layout/MainLayout";
import { useAuth } from "@/contexts/AuthContext";
import { Booking, BookingRequest } from "@/types";
import { getMyClientBookings, getMyBookingRequests } from "@/lib/api";
import SectionList from "@/components/dashboard/SectionList";
import BookingRequestCard from "@/components/dashboard/BookingRequestCard";
import { Spinner } from "@/components/ui";
import DashboardTabs from "@/components/dashboard/DashboardTabs";
import { format } from "date-fns";
import { formatCurrency, formatStatus } from "@/lib/utils";
import Link from "next/link";

export default function ClientDashboardPage() {
  const { user, loading: authLoading } = useAuth();
  const router = useRouter();
  const pathname = usePathname();
  const params = useSearchParams();

  const [bookings, setBookings] = useState<Booking[]>([]);
  const [bookingRequests, setBookingRequests] = useState<BookingRequest[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState("");

  const initialTab = params.get("tab") === "bookings" ? "bookings" : "requests";
  const [activeTab, setActiveTab] = useState<"requests" | "bookings">(initialTab);

  useEffect(() => {
    const params = new URLSearchParams(window.location.search);
    params.set("tab", activeTab);
    router.replace(`${pathname}?${params.toString()}`);
  }, [activeTab, router, pathname]);

  useEffect(() => {
    if (authLoading) return;
    if (!user) {
      router.push(`/login?next=${encodeURIComponent(pathname)}`);
      return;
    }
    if (user.user_type !== "client") {
      router.push("/dashboard/artist");
      return;
    }

    const fetchData = async () => {
      try {
        const [bookingsData, requestsData] = await Promise.all([
          getMyClientBookings(),
          getMyBookingRequests(),
        ]);
        setBookings(bookingsData.data);
        setBookingRequests(requestsData.data);
      } catch (err) {
        console.error("Client dashboard fetch error:", err);
        setError("Failed to load dashboard data. Please try again.");
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [user, authLoading, router, pathname]);

  const upcomingBookings = useMemo(() => {
    const now = new Date().getTime();
    return bookings
      .filter((b) => new Date(b.start_time).getTime() >= now)
      .sort((a, b) => new Date(a.start_time).getTime() - new Date(b.start_time).getTime())
      .slice(0, 5);
  }, [bookings]);

  if (loading) {
    return (
      <MainLayout>
        <div className="p-8 flex justify-center"><Spinner /></div>
      </MainLayout>
    );
  }

  if (error) {
    return (
      <MainLayout>
        <div className="p-8 text-red-600">{error}</div>
      </MainLayout>
    );
  }

  return (
    <MainLayout>
      <div className="max-w-5xl mx-auto px-4 py-8">
        <DashboardTabs activeTab={activeTab} onChange={setActiveTab} showServices={false} />
        <div className="mt-6">
          {activeTab === "requests" && (
            <section>
              <SectionList
                title="Booking Requests"
                data={bookingRequests}
                renderItem={(r) => <BookingRequestCard key={r.id} request={r} onUpdate={() => {}} />}
                emptyState={<span>No requests yet</span>}
              />
            </section>
          )}
          {activeTab === "bookings" && (
            <section>
              <SectionList
                title="Upcoming Bookings"
                data={upcomingBookings}
                emptyState={<span>No bookings yet</span>}
                renderItem={(booking) => (
                  <div key={booking.id} className="bg-white p-4 shadow rounded-lg">
                    <div className="font-medium text-gray-900">
                      {booking.artist?.first_name || "Unknown"} {booking.artist?.last_name || ""}
                    </div>
                    <div className="text-sm text-gray-500">{booking.service?.title || "â€”"}</div>
                    <div className="text-sm text-gray-500">
                      {format(new Date(booking.start_time), "MMM d, yyyy h:mm a")}
                    </div>
                    <div className="mt-2 flex justify-between items-center">
                      <span
                        className={`inline-flex rounded-full px-2 text-xs font-semibold leading-5 ${
                          booking.status === "completed"
                            ? "bg-green-100 text-green-800"
                            : booking.status === "cancelled"
                            ? "bg-red-100 text-red-800"
                            : booking.status === "confirmed"
                            ? "bg-brand-light text-brand-dark"
                            : "bg-yellow-100 text-yellow-800"
                        }`}
                      >
                        {formatStatus(booking.status)}
                      </span>
                      <span className="text-sm text-gray-500">
                        {formatCurrency(Number(booking.total_price))}
                      </span>
                    </div>
                  </div>
                )}
              />
              {bookings.length > upcomingBookings.length && (
                <div className="mt-2">
                  <Link href="/dashboard/client/bookings" className="text-brand-dark hover:underline text-sm">
                    View All Bookings
                  </Link>
                </div>
              )}
            </section>
          )}
        </div>
      </div>
    </MainLayout>
  );
}

