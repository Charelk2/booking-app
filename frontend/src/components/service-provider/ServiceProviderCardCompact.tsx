'use client';
import Image from 'next/image';
import Link, { LinkProps } from 'next/link';
import type { AnchorHTMLAttributes } from 'react';
import { useState } from 'react';
import {
  StarIcon,
} from '@heroicons/react/24/solid';
import clsx from 'clsx';
import { getFullImageUrl } from '@/lib/utils';
import { BREAKPOINT_MD } from '@/lib/breakpoints';

export interface ServiceProviderCardCompactProps
  extends LinkProps,
    Omit<AnchorHTMLAttributes<HTMLAnchorElement>, keyof LinkProps> {
  serviceProviderId: number;
  name: string;
  subtitle?: string;
  imageUrl?: string;
  price?: number;
  rating?: number;
  ratingCount?: number;
  location?: string | null;
  categories?: string[];
  href: string;
}

export default function ServiceProviderCardCompact({
  serviceProviderId,
  name,
  subtitle,
  imageUrl,
  price,
  rating,
  ratingCount,
  location,
  categories,
  href,
  className,
  ...props
}: ServiceProviderCardCompactProps) {
  const [loaded, setLoaded] = useState(false);
  return (
    <Link
      href={href}
      className={clsx(
        'group block rounded-xl overflow-hidden bg-gray-50 hover:shadow-md transition',
        'no-underline hover:no-underline',
        className,
      )}
      {...props}
    >
      <div className="relative aspect-[4/4] bg-gray-100 overflow-hidden">
        {!loaded && (
          <div className="absolute inset-0 animate-pulse bg-gray-200" />
        )}
        {imageUrl ? (
          <Image
            src={imageUrl}
            alt={name}
            fill
            sizes={`(max-width:${BREAKPOINT_MD}px) 50vw, 33vw`}
            className="object-cover w-full h-full group-hover:scale-105 transition-transform"
            onLoad={() => setLoaded(true)}
            onError={(e) => {
              (e.currentTarget as HTMLImageElement).src = getFullImageUrl('/static/default-avatar.svg') as string;
            }}
          />
        ) : (
          <Image
            src={getFullImageUrl('/static/default-avatar.svg') as string}
            alt={name}
            fill
            sizes={`(max-width:${BREAKPOINT_MD}px) 0vw, 33vw`}
            className="object-cover w-full h-full"
            onLoad={() => setLoaded(true)}
          />
        )}
        {rating !== undefined && (
          <span className="absolute top-1 left-1 text-[10px] bg-white/90 rounded-full px-1.5 py-px flex items-center gap-0.5">
            <StarIcon className="h-3 w-3 text-yellow-400" />
            {rating}
            {ratingCount ? (
              <span className="text-gray-500 ml-0.5">({ratingCount})</span>
            ) : null}
          </span>
        )}
        {price !== undefined && (
          <span className="absolute bottom-1 left-1 text-[10px] font-semibold bg-gray-700/60 text-white rounded-full px-1.5 py-px">
            from R{Math.round(price)}
          </span>
        )}
      </div>
      <div className="p-1 space-y-0.5">
        <p className="text-sm font-semibold truncate text-black">{name}</p>
        {subtitle && <p className="text-xs text-gray-600 truncate">{subtitle}</p>}
        {categories && categories.length > 0 && (
          <p className="text-xs text-gray-500 truncate">
            {categories.join(', ')}
          </p>
        )}
        {location && (
          <p className="text-xs text-gray-400 truncate">{location}</p>
        )}
      </div>
    </Link>
  );
}