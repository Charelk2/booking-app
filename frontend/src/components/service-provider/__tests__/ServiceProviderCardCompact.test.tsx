import { createRoot } from 'react-dom/client';
import React from 'react';
import { act } from 'react';
import ServiceProviderCardCompact from '../ServiceProviderCardCompact';

function setup(props = {}) {
  const container = document.createElement('div');
  document.body.appendChild(container);
  const root = createRoot(container);
  const allProps = {
    serviceProviderId: 1,
    name: 'Test',
    href: '/service-providers/1',
    ...props,
  };
  return { container, root, allProps };
}

describe('ServiceProviderCardCompact', () => {
  afterEach(() => {
    document.body.innerHTML = '';
  });

  it('matches snapshot', () => {
    const { container, root, allProps } = setup();
    act(() => {
      root.render(React.createElement(ServiceProviderCardCompact, allProps));
    });
    expect(container.firstChild).toMatchSnapshot();
    act(() => root.unmount());
    container.remove();
  });

  it('ignores subtitle and categories props', () => {
    const { container, root, allProps } = setup({ subtitle: 'DJ', categories: ['Photographer'] });
    act(() => {
      root.render(React.createElement(ServiceProviderCardCompact, allProps));
    });
    expect(container.textContent).not.toContain('DJ');
    expect(container.textContent).not.toContain('Photographer');
    act(() => root.unmount());
    container.remove();
  });

  it('shows only the city when full address is provided', () => {
    const { container, root, allProps } = setup({
      location: '123 Main St, Suburb, Cape Town, Western Cape, South Africa',
    });
    act(() => {
      root.render(React.createElement(ServiceProviderCardCompact, allProps));
    });
    expect(container.textContent).toContain('Cape Town');
    expect(container.textContent).not.toContain('123 Main St');
    expect(container.textContent).not.toContain('Suburb');
    expect(container.textContent).not.toContain('Western Cape');
    act(() => root.unmount());
    container.remove();
  });
});
