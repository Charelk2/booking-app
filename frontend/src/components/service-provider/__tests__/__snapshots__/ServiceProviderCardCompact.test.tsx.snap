// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceProviderCardCompact matches snapshot 1`] = `
<a
  href="/service-providers/1"
>
  <div
    class="relative aspect-[4/4] bg-gray-100 overflow-hidden"
  >
    <div
      class="absolute inset-0 animate-pulse bg-gray-200"
    />
    <img
      alt="Test"
      class="object-cover w-full h-full"
      sizes="(max-width:768px) 0vw, 33vw"
      src="http://localhost:8000/static/default-avatar.svg"
    />
  </div>
  <div
    class="p-1 space-y-0.5"
  >
    <p
      class="text-sm font-semibold truncate text-black"
    >
      Test
    </p>
  </div>
</a>
`;
